### PR TITLE
Micro-optimize type manager

### PR DESCRIPTION
**Pull request by pitrou**
This makes overload resolution around 2x faster.

### CODE REVIEW

**Comment by sklam** [comment]
Out of curiosity, is the new hash table that makes the difference? or, is it de-encapsulation of Type and Rating?  or Both?

**Comment by pitrou** [comment]
It's really a combination of all the changes: the larger hash table, unboxing the Type, improving the selectOverload() logic (i.e. only scan the successful candidates at the end).

**Comment by pitrou** [comment]
Note the new hash() function is modelled on CPython's for tuples; giving us a probably good algorithm.

File: numba/typeconv/typeconv.cpp [hunk 1]
```diff
@@ -184,34 +189,6 @@ int TypeManager::_selectOverload(Type sig[], Type ovsigs[], int &selected,
     return matchcount;
 }
 
-// ----- Ratings -----
-Rating::Rating() :promote(0), safe_convert(0), unsafe_convert(0) { }
-
-void Rating::bad() {
```
Comment (sklam) [inline numba/typeconv/typeconv.cpp]: Can you remove the declaration in the hpp?

Comment (pitrou) [inline numba/typeconv/typeconv.cpp]: No, the Rating type is still used, it's just moved around :)
(I was assuming that maybe the compiler wasn't able to inline something that's defined after its use, but I may be wrong about that)

Comment (sklam) [inline numba/typeconv/typeconv.cpp]: I meant Rating::bad() at https://github.com/pitrou/numba/blob/faster_typemanager/numba/typeconv/typeconv.hpp#L55

Comment (pitrou) [inline numba/typeconv/typeconv.cpp]: Oh, yes, sorry. I'll remove it.

**Comment by pitrou** [comment]
I've removed the declaration.

**Pull request merged**